### PR TITLE
Create courses functionality for creating, listing, and joining courses

### PR DESCRIPTION
**Pull request by SamHaggans**
This creates six new commands with the bot:

1.     /course create [name] [crn] [description] - Creates a new course and associated role in discord. Requires the Manage Roles permission.
2.     /course remove [role] - Removes a given course and the associated role. Requires the Manage Roles permission.
3.     /course join [role] - Joins a course and assigns the appropriate course role to the user.
4.     /course leave [role] - Leaves a class and removes the appropriate course role from the user.
5.     /course list - Lists the available courses that can be joined in a message only visible to the requested user.
6.     /course roster [role] - Lists the members currently in that course.

### CODE REVIEW

File: src/commands/course/subcommands/_roster.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import * as Discord from '../../../Discord';
+import SubCommand from '../../../SubCommand';
+import CommandOption from '../../../CommandOption';
+import { CommandOptionType } from '../../../Discord';
+import { courseExists, getCourse, getCourseMembers } from '../_common';
+
+export const roster = new SubCommand({
+  name: 'roster',
+  displayName: 'Returns a roster of members in a course',
```
Comment (jaysella) [inline src/commands/course/subcommands/_roster.ts]:
```suggestion
  displayName: 'Course Roster',
```

**Review by jaysella — CHANGES_REQUESTED** [review]
This looks great! I left one requested change to switch the `displayName` for `remove`. I also committed documentation for `COMMANDS.md`.

**Review by thederekkaplan — CHANGES_REQUESTED** [review]
As discussed in the Discord:
- [x] Change from using CRNs to course numbers
- [x] Use `_id` to store in database

**Review by thederekkaplan — APPROVED** [review]

**Review by aiyan — CHANGES_REQUESTED** [review]
Are we going to change the number system to the subject code + 4-digit number?

**Review by jaysella — APPROVED** [review]

**Review by aiyan — APPROVED** [review]

File: src/commands/course/_common.ts [hunk 2]
```diff
@@ -1,19 +1,48 @@
 import { Cursor, Collection as MongoCollection } from 'mongodb';
 import { Collection } from '../../database';
 import ExecutionContext from '../../ExecutionContext';
+import { bold } from '../../format';
 
 export type Course = {
-  _id: string;
+  _id: number;
```
Comment (thederekkaplan) [inline src/commands/course/_common.ts]: If we're using just the number and not also the course prefix as _id, then it'll create a conflict between courses in different subjects with the same number.

**Review by thederekkaplan — CHANGES_REQUESTED** [review]

File: src/ExecutionContext.ts [hunk 3]
```diff
@@ -117,6 +117,25 @@ class ExecutionContext {
...
+    const data: Discord.InteractionCommandCallbackData = { embeds };
+    if (ephemeral) {
+      data.flags = 64;
+    }
+
+    return this.respond({
+      type: Discord.InteractionCallbackType.ChannelMessageWithSource,
+      data,
+    });
+  }
```
Comment (jaysella) [inline src/ExecutionContext.ts]: These are missing documentation `/** */`

File: src/commands/course/_common.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+export const deleteCourse = async (
+  ctx: ExecutionContext,
+  guildId: string,
+  courseInfo: Course,
+): Promise<void> => {
+  const db = await ctx.db.getDb(guildId);
+
+  await db
+    .collection(Collection.COURSES)
+    .deleteOne({ roleId: courseInfo.roleId });
```
Comment (jaysella) [inline src/commands/course/_common.ts]: Can `coursesCollection()` be used here instead?
```suggestion
  coursesCollection(ctx, guildId).deleteOne({ roleId: courseInfo.roleId }));
```

File: src/commands/course/subcommands/_list.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }));
+
+    await ctx.respondWithEmbed(
+      {
+        type: Discord.EmbedType.RICH,
+        title: 'Course List',
+        fields,
+      },
+      true,
+    );
```
Comment (jaysella) [inline src/commands/course/subcommands/_list.ts]: To match what was done with `resondSilently()` should a `respondWithSilentEmbed()` be created?

File: src/commands/course/subcommands/_roster.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    return ctx.respondWithEmbed(
+      {
+        type: Discord.EmbedType.RICH,
+        title: `Members in ${boldCourse(course)}`,
+        description: memberList,
+      },
+      true,
+    );
```
Comment (jaysella) [inline src/commands/course/subcommands/_roster.ts]: `respondWithSilentEmbed()`

**Review by jaysella — CHANGES_REQUESTED** [review]

**Review by thederekkaplan — APPROVED** [review]

**Review by jaysella — APPROVED** [review]

**Pull request merged**